cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/DBSource.java

```diff
@@ -124,6 +131,48 @@ private String getJDBCPluginId() {
     return String.format("%s.%s.%s", "source", dbSourceConfig.jdbcPluginType, dbSourceConfig.jdbcPluginName);
   }
 
+  private void ensureValidConnection(Class<? extends Driver> jdbcDriverClass) {
+    try {
+      ensureJDBCDriverIsAvailable(jdbcDriverClass);
+    } catch (Exception e) {
+      throw new IllegalArgumentException("could not ensure JDBCDriver is available", e);
+    }
+    try (Connection connection = createConnection()
+    ) {
+      assert(connection.isValid(0));
+    } catch (SQLException e) {
+      throw new IllegalArgumentException("SQL Exception thrown when trying to connect to driver", e);
+    } finally {
+      DBUtils.cleanup(jdbcDriverClass);
+    }
+  }
+
+  /*
+   * Throws InstantiationException or IllegalAccessException if it can't get access to the jdbcDriverClass
+   * In either case, the connection to the driver failed
+   */
+  private void ensureJDBCDriverIsAvailable(Class<? extends Driver> jdbcDriverClass) throws Exception {
+    try {
+      DriverManager.getDriver(dbSourceConfig.connectionString);
+    } catch (SQLException e) {
+      // Driver not found. We will try to register it with the DriverManager.
+      LOG.debug("Plugin Type: {} and Plugin Name: {}; Driver Class: {} not found. Registering JDBC driver via shim {} ",
+                dbSourceConfig.jdbcPluginType, dbSourceConfig.jdbcPluginName, jdbcDriverClass.getName(),
+                JDBCDriverShim.class.getName());
+      if (driverShim == null) {
+        driverShim = new JDBCDriverShim(jdbcDriverClass.newInstance());
+      }
+      DBUtils.deregisterAllDrivers(jdbcDriverClass);
```

Comment:
So the purpose of this is to clean up default drivers that get added when you create a new instance of the ``jdbcDriverClass``. I think you could move it inside the above ``if`` statement, where you create the new instance. 

It would also be good to confirm this. You can run the ``BatchETLDBAdapterTest`` in debug mode, place a breakpoint in the ``deregisterDriver`` method in ``DriverManager.java`` and inspect how the ``registeredDrivers`` list gets modified when this code is executed. You should ensure that that list does not get filled up with new instances of ``jdbcDriverClass``, without getting cleaned up. 

I can help you with this tomorrow too, if you are unsure about